cpp/tools/parquet/parquet_dump_footer.cc

```diff
@@ -0,0 +1,246 @@
+    des.DeserializeMessage(reinterpret_cast<const uint8_t*>(data), &len, obj);
+    return true;
+  } catch (const std::exception& e) {
+    std::cerr << "Failed to deserialize: " << e.what() << "\n";
+    return false;
+  }
+}
+
+template <typename T>
+bool Serialize(const T& obj, std::string* out) {
+  parquet::ThriftSerializer ser(/*initial_buffer_size=*/10 << 20);
+  try {
+    ser.SerializeToString(&obj, out);
+    return true;
+  } catch (const std::exception& e) {
+    std::cerr << "Failed to serialize: " << e.what() << "\n";
+    return false;
+  }
+}
+
+// Replace the contents of s with random data of the same length.
+void Scrub(std::string* s) {
+  static char pool[4096];
+  static std::mt19937 rng(std::random_device {}());
+  static const bool kPoolInit = [] {
+    std::uniform_int_distribution<> caps(65, 90);
+    for (size_t i = 0; i < sizeof(pool); i++) pool[i] = caps(rng);
+    return true;
+  }();
+  (void)kPoolInit;
+
+  const size_t n = s->size();
+  s->clear();
+  while (s->size() < n) {
+    size_t m = std::min(n, sizeof(pool) / 2);
+    std::uniform_int_distribution<> start(0, sizeof(pool) / 2);
+    s->append(&pool[start(rng)], m);
+  }
+}
+
+void Scrub(parquet::format::FileMetaData* md) {
+  for (auto& s : md->schema) {
+    Scrub(&s.name);
+  }
+  for (auto& r : md->row_groups) {
+    for (auto& c : r.columns) {
+      Scrub(&c.file_path);
+      if (c.__isset.meta_data) {
+        auto& m = c.meta_data;
+        for (auto& p : m.path_in_schema) Scrub(&p);
+        for (auto& kv : m.key_value_metadata) {
+          Scrub(&kv.key);
+          Scrub(&kv.value);
+        }
+        Scrub(&m.statistics.max_value);
+        Scrub(&m.statistics.min_value);
+        Scrub(&m.statistics.min);
+        Scrub(&m.statistics.max);
+      }
+
+      if (c.crypto_metadata.__isset.ENCRYPTION_WITH_COLUMN_KEY) {
+        auto& m = c.crypto_metadata.ENCRYPTION_WITH_COLUMN_KEY;
+        for (auto& p : m.path_in_schema) Scrub(&p);
+        Scrub(&m.key_metadata);
+      }
+      Scrub(&c.encrypted_column_metadata);
+    }
+  }
+  for (auto& kv : md->key_value_metadata) {
+    Scrub(&kv.key);
+    Scrub(&kv.value);
+  }
+  Scrub(&md->footer_signing_key_metadata);
+}
+
+// Returns:
+// - 0 on success
+// - -1 on error
+// - the size of the footer if tail is too small
+int64_t ParseFooter(const std::string& tail, parquet::format::FileMetaData* md) {
+  if (tail.size() > std::numeric_limits<int32_t>::max()) return -1;
+
+  const char* p = tail.data();
+  const int32_t n = static_cast<int32_t>(tail.size());
+  int32_t len = ReadLE32(p + n - 8);
+  if (len > n - 8) return len;
+
+  if (!Deserialize(tail.data() + n - 8 - len, len, md)) return -1;
+  return 0;
+}
+}  // namespace
+
+int main(int argc, char** argv) {
+  bool help = false;
+  bool scrub = true;
+  bool json = false;
+  std::string in;
+  std::string out;
+  for (int i = 1; i < argc; i++) {
+    char* arg = argv[i];
+    help |= !std::strcmp(arg, "-h") || !std::strcmp(arg, "--help");
```

Comment:
Google's Gflags library is included in the [build system](https://github.com/apache/arrow/blob/84df3438e39b470ec9bfe77e682a14bda07b0921/cpp/src/arrow/flight/flight_benchmark.cc#L462) I think (but might be test only it might be nice to use that instead of manual parsing)

Reply:
It looks like other binaries take this manual parsing approach I'm not sure if this was done because flags was originally not included or test only.

Reply:
I tried but I get missing symbols:

```
/usr/bin/ld: tools/parquet/CMakeFiles/parquet-dump-footer.dir/parquet_dump_footer.cc.o: in function `main':
parquet_dump_footer.cc:(.text.startup+0x54): undefined reference to `google::SetUsageMessage(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/usr/bin/ld: parquet_dump_footer.cc:(.text.startup+0x87): undefined reference to `google::ParseCommandLineFlags(int*, char***, bool)'
/usr/bin/ld: tools/parquet/CMakeFiles/parquet-dump-footer.dir/parquet_dump_footer.cc.o: in function `_GLOBAL__sub_I__ZN3fLB14FLAGS_no_scrubE':
parquet_dump_footer.cc:(.text.startup+0x12f7): undefined reference to `google::FlagRegisterer::FlagRegisterer<bool>(char const*, char const*, char const*, bool*, bool*)'
/usr/bin/ld: parquet_dump_footer.cc:(.text.startup+0x1326): undefined reference to `google::FlagRegisterer::FlagRegisterer<bool>(char const*, char const*, char const*, bool*, bool*)'
/usr/bin/ld: parquet_dump_footer.cc:(.text.startup+0x138c): undefined reference to `google::FlagRegisterer::FlagRegisterer<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(char const*, char const*, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)'
/usr/bin/ld: parquet_dump_footer.cc:(.text.startup+0x1425): undefined reference to `google::FlagRegisterer::FlagRegisterer<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(char const*, char const*, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)'
```

Reply:
OK, I guess manual parsing might be easier.  Unless someone better acquanted with CMake can help out.

Reply:
`target_link_libraries(parquet-dump-footer thrift::thrift ${ARROW_LIBRARIES} ${GFLAGS_LIBRARIES})` may work.